### PR TITLE
hotfix compilation windows

### PR DESCRIPTION
**Pull request by philbucher**
Moved a constructor (back) from protected to private, since this gave an error when compiled in windows
@dbaumgaertner does it compile now
@roigcarlo I found the error, however I cannot say why it complained in windows

### CODE REVIEW

**Comment by roigcarlo** [comment]
@philbucher Do you have the error message?

**Comment by philbucher** [comment]
This is the error message of @dbaumgaertner 
```
"C:\Kratos\cmake_build\INSTALL.vcxproj" (Standardziel) (1) ->

"C:\Kratos\cmake_build\ALL_BUILD.vcxproj" (Standardziel) (3) ->

"C:\Kratos\cmake_build\applications\MappingApplication\KratosMappingApplication.vcxproj" (Standardziel) (11) ->

(ClCompile Ziel) ->

  C:\Kratos\kratos\includes/serializer.h(227): error C2248: 'Kratos::InterfaceObject::InterfaceObject': cannot access protected member declared in class 'Kratos::InterfaceObject' [C:\Kratos\cma

ke_build\applications\MappingApplication\KratosMappingApplication.vcxproj]
```


This is on another system, @peterjwilson , with which I could figure it out :
```
1>C:\KratosGithub\Kratos\kratos\includes/serializer.h(227): error C2248: 'Kratos::InterfaceObject::InterfaceObject': cannot access protected member declared in class 'Kratos::InterfaceObject'

1>  C:\KratosGithub\Kratos\applications\MappingApplication\custom_utilities/interface_object.h(256): note: see declaration of 'Kratos::InterfaceObject::InterfaceObject'

1>  C:\KratosGithub\Kratos\applications\MappingApplication\custom_utilities/interface_object.h(63): note: see declaration of 'Kratos::InterfaceObject
'```

**Review by roigcarlo — APPROVED** [review]
Oh, the serializer one. I remember. No clues about that one yet, but changes are ok. :+1:

**Comment by pooyan-dadvand** [comment]
I don't understand why the friend cannot see the protected member. I think there is another issue....

**Comment by philbucher** [comment]
@pooyan-dadvand it is weird, the serializer complains even though I don't use it at all in the application

**Comment by pooyan-dadvand** [comment]
The problem comes from the fact that you have not made `Serializer` a friend of your class and its base class point is serialized.

**Comment by philbucher** [comment]
Thanks @pooyan-dadvand 
this means that all classes that derive from `Point` must have the Serializer as friend?

**Comment by pooyan-dadvand** [comment]
Yes, because serializer tries to call the correct derived class constructor.

**Comment by philbucher** [comment]
I added the serializer
@peterjwilson and @dbaumgaertner could you please compile it under windows and see if it works?

**Comment by peterjwilson** [comment]
Yep it compiles all good 👍 @philbucher

**Comment by dbaumgaertner** [comment]
@philbucher: It also compiles on my computer under windows now.

File: applications/MappingApplication/custom_utilities/interface_geometry_object.h [hunk 1]
```diff
@@ -389,6 +389,16 @@ class InterfaceGeometryObject : public InterfaceObject
...
     ///@{
 
+        
+    ///@}
+    ///@name Serialization
+    ///@{
+
+    friend class Serializer;
+    
+    virtual void save(Serializer& rSerializer) const {}
```
Comment (pooyan-dadvand) [inline applications/MappingApplication/custom_utilities/interface_geometry_object.h]: You should call the base class `save` class as described in the [old wiki](http://kratos-wiki.cimne.upc.edu/index.php/How_to_use_Serialization)

File: applications/MappingApplication/custom_utilities/interface_geometry_object.h [hunk 2]
```diff
@@ -389,6 +389,16 @@ class InterfaceGeometryObject : public InterfaceObject
...
+    virtual void load(Serializer& rSerializer) {}
```
Comment (pooyan-dadvand) [inline applications/MappingApplication/custom_utilities/interface_geometry_object.h]: Same for the `load` method

File: applications/MappingApplication/custom_utilities/interface_node.h [hunk 3]
```diff
@@ -236,6 +236,16 @@ class InterfaceNode : public InterfaceObject
...
     ///@{
 
+        
+    ///@}
+    ///@name Serialization
+    ///@{
+
+    friend class Serializer;
+    
+    virtual void save(Serializer& rSerializer) const {}
```
Comment (pooyan-dadvand) [inline applications/MappingApplication/custom_utilities/interface_node.h]: Again.

File: applications/MappingApplication/custom_utilities/interface_node.h [hunk 4]
```diff
@@ -236,6 +236,16 @@ class InterfaceNode : public InterfaceObject
...
+    virtual void load(Serializer& rSerializer) {}
```
Comment (pooyan-dadvand) [inline applications/MappingApplication/custom_utilities/interface_node.h]: Again

**Review by pooyan-dadvand — CHANGES_REQUESTED** [review]
you should call the base class save and load methods using corresponding macro as described in the wiki. Meanwhile would you please move the serializer how to from the old wiki to the new one?

**Review by dbaumgaertner — APPROVED** [review]

**Comment by philbucher** [comment]
@pooyan-dadvand I Implemented your requests, can you please have a look?
I have some questions regarding the serializer, after reading the wiki:
The class structure is the following: `InterfaceNode` and `InterfaceGeometryObject` derive from `InterfaceObject`, which itself derives from `Point<3>`
1. Do I have to add the three Objects in the `mapping_application.h` and `mapping_application.cpp` as described in the wiki? I tried this but ofc it is not working since none of these classes have a public default constructor (which I don't really want to add...)
2. `InterfaceObject` has a protected default constructor, but should I add one (if yes, then in private or protected) to the deriving classes, `InterfaceNode` and `InterfaceGeometryObject`? I don't intend to further derive from these classes

Btw at least for now I don't intend to use Serialization for the `MappingApplication`. Also none of the above mentioned objects is exposed to python. In fact, they will be superfluous and removed once the mpi-search and the global pointers are there

Also I [ported the serializer-wiki](https://github.com/KratosMultiphysics/Kratos/wiki/How-to-use-Serialization). Please let me know if it is ok

**Comment by pooyan-dadvand** [comment]
First of all thanks for migrating the wiki! 

You should add default constructor to ones they don't have any (better private). And what I have noticed is that you have not save and loading your members.

**Comment by philbucher** [comment]
At the end of [this section](https://github.com/KratosMultiphysics/Kratos/wiki/How-to-use-Serialization#preparing-a-class-for-serialization) it says that I should add default constructor calls to the `mapping_application.h` and `mapping_application.cpp`. These would call the default constructor in the **public** section, which do not exist.
Thats why I was asking where to put them. If I put them in private this won't compile.
So far I haven't added anything to theses files.

I don't add the members bcs I don't intend to use Serialization (for now), also the rest of the application is not prepared

**Comment by jcotela** [comment]
@philbucher I don't think you need a default constructor to add the objects to `mapping_application.h/.cpp`, but you need a reference to an object of the correct type to call Serializer::Register. You just need to be able to use any public constructor to initialize a dummy instance of your class.

**Comment by philbucher** [comment]
Hi Jordi,
Thx for your help.
However the problem remains:
As soon as I add the members in `mapping_application.h` it expects a public default constructor.
I (am quite sure I) did now exactly what is written in the wiki, but this doesn't compile:
```
In file included from /home/philippb/software/Kratos/applications/MappingApplication/mapping_application.h:34:0,
                 from /home/philippb/software/Kratos/applications/MappingApplication/mapping_application.cpp:24:
/home/philippb/software/Kratos/applications/MappingApplication/custom_utilities/interface_object.h: In constructor ‘Kratos::KratosMappingApplication::KratosMappingApplication()’:
/home/philippb/software/Kratos/applications/MappingApplication/custom_utilities/interface_object.h:257:5: error: ‘Kratos::InterfaceObject::InterfaceObject()’ is protected
     InterfaceObject() : Point<3>(0.0f, 0.0f, 0.0f)
     ^
/home/philippb/software/Kratos/applications/MappingApplication/mapping_application.cpp:38:52: error: within this context
 KratosMappingApplication::KratosMappingApplication() { }
                                                    ^
In file included from /home/philippb/software/Kratos/applications/MappingApplication/mapping_application.h:35:0,
                 from /home/philippb/software/Kratos/applications/MappingApplication/mapping_application.cpp:24:
/home/philippb/software/Kratos/applications/MappingApplication/custom_utilities/interface_node.h:227:5: error: ‘Kratos::InterfaceNode::InterfaceNode()’ is private
     InterfaceNode() : InterfaceObject()
     ^
/home/philippb/software/Kratos/applications/MappingApplication/mapping_application.cpp:38:52: error: within this context
 KratosMappingApplication::KratosMappingApplication() { }
                                                    ^
In file included from /home/philippb/software/Kratos/applications/MappingApplication/mapping_application.h:36:0,
                 from /home/philippb/software/Kratos/applications/MappingApplication/mapping_application.cpp:24:
/home/philippb/software/Kratos/applications/MappingApplication/custom_utilities/interface_geometry_object.h:351:5: error: ‘Kratos::InterfaceGeometryObject::InterfaceGeometryObject()’ is private
     InterfaceGeometryObject() : InterfaceObject()
     ^
/home/philippb/software/Kratos/applications/MappingApplication/mapping_application.cpp:38:52: error: within this context
 KratosMappingApplication::KratosMappingApplication() { }
```

**Comment by jcotela** [comment]
@philbucher The problem is in attempting to call the default constructor in `mapping_application.cpp`, which the compiler rightly complains about. Would it be acceptable to call the public constructor with dummy arguments there? Something like
```
KratosMappingApplication::KratosMappingApplication():
    mInterfaceObject(0.0f,0.0f,0.0f);
    ...
{}
```

**Comment by philbucher** [comment]
@jcotela I did these modifications, now it compiles.
However now  `InterfaceNode` and `InterfaceGeometryObject` have public default constructors, which I would have liked to avoid.
But it seems like there is no way around it if I am not mistaken?

**Comment by jcotela** [comment]
Well, the two classes you are referring to already have (non-default) public constructors. If you don't want to add default constructors, which I can understand, the alternative is to (ab)use the existing ones. You'll need to create a dummy node/geometry on the spot, but that should not be a problem.

**Comment by jcotela** [comment]
Actually... I see now that the existing constructors take node/geometry *references*. You can still survive that by defining those as class members, but it is getting uglier, I agree.

**Comment by philbucher** [comment]
I think then I leave it is as is, with the new constructors added.
in this case I will ask the windows guys once again once @pooyan-dadvand has checked this
Thanks @jcotela

**Comment by pooyan-dadvand** [comment]
It is ok to call the parent constructors with dummy arguments from a private constructor. I prefer this to adding a default constructor to the parent class. Nevertheless you still missing the implementation of the save and load of each class member.

**Comment by philbucher** [comment]
Added error messages to the serializer methods, as we discussed
@pooyan-dadvand is it ok now?

**Comment by philbucher** [comment]
@dbaumgaertner could you please try to compile in in windows to make sure it works?

**Comment by pooyan-dadvand** [comment]
So finally are you registering this objects?

**Comment by philbucher** [comment]
@pooyan-dadvand you are right, I had forgotten to remove the registering

**Review by pooyan-dadvand — APPROVED** [review]

**Comment by pooyan-dadvand** [comment]
:+1:

**Pull request merged**